src/herder/TransactionQueue.cpp

```diff
@@ -602,148 +489,92 @@ TransactionQueue::tryAdd(TransactionFrameBasePtr tx, bool submittedFromSelf)
+
+    prepareDropTransaction(stateIter->second);
 
-    // Actually erase the transactions to be dropped.
-    stateIter->second.mTransactions.erase(begin, end);
+    // Actually erase the transaction to be dropped.
+    stateIter->second.mTransaction.reset();
 
     // If the queue for stateIter is now empty, then (1) erase it if it is not
     // the fee-source for some other transaction or (2) reset the age otherwise.
-    if (stateIter->second.mTransactions.empty())
+    if (stateIter->second.mTotalFees == 0)
     {
-        if (stateIter->second.mTotalFees == 0)
-        {
-            mAccountStates.erase(stateIter);
-        }
-        else
-        {
-            stateIter->second.mAge = 0;
-        }
+        mAccountStates.erase(stateIter);
+    }
+    else
+    {
+        stateIter->second.mAge = 0;
     }
 }
 
 void
 TransactionQueue::removeApplied(Transactions const& appliedTxs)
 {
     ZoneScoped;
-    // Find the highest sequence number that was applied for each source account
-    std::map<AccountID, int64_t> seqByAccount;
-    UnorderedSet<Hash> appliedHashes;
-    appliedHashes.reserve(appliedTxs.size());
-    for (auto const& tx : appliedTxs)
-    {
-        auto& seq = seqByAccount[tx->getSourceID()];
-        seq = std::max(seq, tx->getSeqNum());
-        appliedHashes.emplace(tx->getFullHash());
-    }
 
     auto now = mApp.getClock().now();
-    for (auto const& kv : seqByAccount)
+    for (auto const& appliedTx : appliedTxs)
     {
         // If the source account is not in mAccountStates, then it has no
         // transactions in the queue so there is nothing to do
-        auto stateIter = mAccountStates.find(kv.first);
+        auto stateIter = mAccountStates.find(appliedTx->getSourceID());
         if (stateIter != mAccountStates.end())
         {
             // If there are no transactions in the queue for this source
             // account, then there is nothing to do
-            auto& transactions = stateIter->second.mTransactions;
-            if (!transactions.empty())
+            auto const& transaction = stateIter->second.mTransaction;
+            if (transaction)
             {
-                // If the sequence number of the first transaction is greater
-                // than the highest applied sequence number for this source
-                // account, then there is nothing to do because sequence numbers
-                // are monotonic (this shouldn't happen)
-                if (transactions.front().mTx->getSeqNum() <= kv.second)
+                // We care about matching the sequence number rather than
+                // the hash, because any transaction with a sequence number
+                // less-than-or-equal to the highest applied sequence number
+                // for this source account has either (1) been applied, or
+                // (2) become invalid.
+                if (transaction->mTx->getSeqNum() <= appliedTx->getSeqNum())
```

Comment:
Do this condition and comment still make sense? IIUC we should only have one tx/account now, so seq num should be strictly the same, right?

Reply:
yeah, the condition is up-to-date: we can't make assumptions about sequence numbers here, because what the node applies is not guaranteed to match what's in the queue. So here if we applied a tx with seqnum N, we need to ban all earlier txs as well.